###### Introduction to Network Analysis 2023/24 (x)

## Node mixing by (not) degree, graphlet degrees

### III. Node mixing by _not_ degree



In [ ]:
import networkx as nx
import utils

1. **(homework)** Study node mixing in networks by some property _other_ than node degree $k$. This can be either some structural property of nodes (e.g., node clustering coefficient $C$ or $C^\mu$) or external information associated with each node (e.g., sociological partitioning of nodes in social network or traffic loads in highway networks).

In [ ]:
def fvalue_parser(label, value):
	mdata = {"label": label}
	if value is not None and len(value.strip()) > 0:
		mdata["value"] = float(value)
	else:
		mdata["value"] = 0

	return mdata

In [ ]:
print("{:>21s} | {:^7s} {:^7s}".format('Graph', 'r(d)', 'r(n)'))

for name in ["karate_club", "southern_women", "cdn_java", "cdn_jung", "wikileaks", "highways"]:
    G = utils.read_pajek(name, label_parser=fvalue_parser)
    rd = nx.attribute_assortativity_coefficient(G, 'value')
    rn = nx.numeric_assortativity_coefficient(G, 'value')

    print("{:>21s} | {:^7.3f} {:^7.3f}".format("'" + G.name + "'", rd, rn))
print()


                Graph |  r(d)    r(n)  
        'karate_club' |  0.743   0.743 
     'southern_women' | -0.603  -0.829 
           'cdn_java' |  0.272   0.570 
           'cdn_jung' |  0.456   0.567 
          'wikileaks' |  0.834   0.764 
           'highways' |  0.112   0.600 



In example above we are testing out the asortitivity/disassortitivity of the value attribute of our nodes. In example of karate_club, each node has label of 1 or 2, depending on which karate club people started attending after the original club split. In this way we will determinate how likely the trainees from club 1 are connected to people from club 1 or 2 and vice versa.

|    Graph        |  r(d)   |   r(n)   |
| ----------------| ------- | -------- |
|'karate_club'    |  0.743   |  0.743  |
|'southern_women' | -0.603   | -0.829  |
|'cdn_java'       |  0.272   |  0.570  |
|'cdn_jung'       |  0.456   |  0.567  |
|'wikileaks'      |  0.834   |  0.764  |
|'highways'       |  0.112   |  0.600  |